### ISSUE TITLE
local network exclude

### ISSUE THREAD

**Issue by domhaas** [schwabe/ics-openvpn#1202]
To make issues more manageable, I would appreciate it if you fill out the following details as applicable:

# General information
1. Android Version: Android 10
2. Android Vendor/Custom ROM: GrapheneOS (same Issue also on Google pixel3a stock)
3. Device: Pixel3a
4. Version of the app: latest
 
# Description of the issue

When I'm on mobile, my android should use VPN to connect to my home-network (this works perfectly)- when I'm connected with local (Wifi) the endpoints should be connected directly - this goes also through the tunnel.

The logfile says:
Routes excluded: (...)/64

So there is only an ipv6 listened, not the ipv4 what is all about.

Got a similar [issue](https://github.com/schwabe/ics-openvpn/issues/1130) to not go through VPN on a local network.

Any ideas what is the issue? I've disabled ipv6 on my router, but anyhow Android is going to create an ipv6 local-address.

Thank you in advance.

**Comment by schwabe** [comment]
Have you enabled the option to exclude local networks?

**Comment by domhaas** [comment]
> Have you enabled the option to exclude local networks?

Thanks for the quick reply.

Yes, the option to exclude local networks is activated. I've also tested to put my local-network (192.168.10.0/24) manually to the exclude list - this works, but then this network is of course not working when not on wifi.

I've had this issue some weeks ago also on Stock-Android 10 on my pixel 3a and again since I've switched to GrapheneOS a week ago.
I'm going to check if it has something to do with AOSP and some restriction. Also I'm going to prepare you some logs.

Thank you in advance.

edit: I'm going to create a test-app to check the network-detection on this base:
[https://github.com/schwabe/ics-openvpn/blob/master/main/src/main/java/de/blinkt/openvpn/core/NetworkUtils.java#L19](https://github.com/schwabe/ics-openvpn/blob/master/main/src/main/java/de/blinkt/openvpn/core/NetworkUtils.java#L19)

**Comment by schwabe** [comment]
hm strange. Maybe the function to detect local network is not working as expected. Maybe you wifi interface has a very strange name or so. Do you have adb setup and could post a `ip addr` when on your wifi?

**Comment by schwabe** [comment]
The app uses this: https://developer.android.com/reference/android/net/ConnectivityManager#getAllNetworks() API to query all interfaces. 

It works fine with my Pixel devices. So I wonder what makes the 3a special here.

**Comment by domhaas** [comment]
> hm strange. Maybe the function to detect local network is not working as expected. Maybe you wifi interface has a very strange name or so. Do you have adb setup and could post a `ip addr` when on your wifi?

Nothing special, just a normal name.
adb ip addr:

```
30: wlan0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc mq state UP group default qlen 3000
    link/ether <hidden> brd ff:ff:ff:ff:ff:ff
    inet 192.168.10.40/24 brd 192.168.10.255 scope global wlan0
       valid_lft forever preferred_lft forever
    inet6 <hidden>/64 scope link stable-privacy 
       valid_lft forever preferred_lft forever
31: wlan1: <BROADCAST,MULTICAST> mtu 1500 qdisc noop state DOWN group default qlen 3000
    link/ether <hidden> brd ff:ff:ff:ff:ff:ff
```

**Comment by schwabe** [comment]
>  I'm going to create a test-app to check the network-detection on this base: https://github.com/schwabe/ics-openvpn/blob/master/main/src/main/java/de/blinkt/openvpn/core/NetworkUtils.java#L19

Yeah that would be super helpful. Because all I can think of currently is to add more logging to that function to figure out if Android is not reporting your WiFI or if I screwed up. You can also skip the `} else {` part for older Android. That one requires the use of JNI.

**Comment by domhaas** [comment]
> > I'm going to create a test-app to check the network-detection on this base: https://github.com/schwabe/ics-openvpn/blob/master/main/src/main/java/de/blinkt/openvpn/core/NetworkUtils.java#L19
> 
> Yeah that would be super helpful. Because all I can think of currently is to add more logging to that function to figure out if Android is not reporting your WiFI or if I screwed up. You can also skip the `} else {` part for older Android. That one requires the use of JNI.

So I've justed created a minimal-app and using your NetworkUtils.getLocalNetworks()
```
(...)
Vector<String> test = NetworkUtils.getLocalNetworks(this, false);
Log.i("getLocalNetworks()", test.toString());
(...)
```
and it works correctly:
`com.example.myapplication I/getLocalNetworks(): [192.168.10.40/24]`

So the method is actually working on my pixel3a, the problem seems to be elsewhere. Any ideas?

**Comment by domhaas** [comment]
> > > I'm going to create a test-app to check the network-detection on this base: https://github.com/schwabe/ics-openvpn/blob/master/main/src/main/java/de/blinkt/openvpn/core/NetworkUtils.java#L19
> > 
> > 
> > Yeah that would be super helpful. Because all I can think of currently is to add more logging to that function to figure out if Android is not reporting your WiFI or if I screwed up. You can also skip the `} else {` part for older Android. That one requires the use of JNI.
> 
> So I've justed created a minimal-app and using your NetworkUtils.getLocalNetworks()
> 
> ```
> (...)
> Vector<String> test = NetworkUtils.getLocalNetworks(this, false);
> Log.i("getLocalNetworks()", test.toString());
> (...)
> ```
> 
> and it works correctly:
> `com.example.myapplication I/getLocalNetworks(): [192.168.10.40/24]`
> 
> So the method is actually working on my pixel3a, the problem seems to be elsewhere. Any ideas?

Oh - 192.168.10.40/24 is wrong:
should be 192.168.10.0/24 - so that seems the issue.

Also I'm getting this inside logs:
`exception parsing IPv4 route: [route] [192.168.10.40/24] (...) tun_prop_error: route is not canonical`

**Comment by schwabe** [comment]
ah yeah, posting the whole log earlier would been helpful. Can you still post the log? I want to understand how that route ends up to tunbuilder but not show in excluded routes. That sounds extremely weird. I expect you have a corner case of overlapping routes from VPN and local routes in which case the app prefer the routes the routes from the VPN but additionally your route from the VPN server is not in canonical form.

**Comment by domhaas** [comment]
> ah yeah, posting the whole log earlier would been helpful. Can you still post the log? I want to understand how that route ends up to tunbuilder but not show in excluded routes. That sounds extremely weird. I expect you have a corner case of overlapping routes from VPN and local routes in which case the app prefer the routes the routes from the VPN but additionally your route from the VPN server is not in canonical form.

Sorry for the late log. Private-Infos marked with <#HIDDEN#>

```
2020-06-06 18:28:38 F-Droid built and signed version 0.7.15 running on Android Pixel 3a (sargo), Android 10 (QQ3A.200605.002) API 29, ABI arm64-v8a, (Android/aosp_sargo/sargo:10/QQ3A.200605.002/2020.06.02.02:user/release-keys)
2020-06-06 18:28:38 Building configuration…
2020-06-06 18:28:39 Network Status: CONNECTED  to WIFI 
2020-06-06 18:28:39 Debug state info: CONNECTED  to WIFI , pause: userPause, shouldbeconnected: true, network: SHOULDBECONNECTED 
2020-06-06 18:28:39 Debug state info: CONNECTED  to WIFI , pause: userPause, shouldbeconnected: true, network: SHOULDBECONNECTED 
2020-06-06 18:28:39 OpenVPN core 3.6_git:master(icsopenvpn/v0.7.14-0-g5391351a) android arm64 64-bit
2020-06-06 18:28:39 Copyright (C) 2012-2017 OpenVPN Inc. All rights reserved.
2020-06-06 18:28:39 OpenVPN core 3.6_git:master(icsopenvpn/v0.7.14-0-g5391351a) android arm64 64-bit
2020-06-06 18:28:39 Frame=512/2048/512 mssfix-ctrl=1250
2020-06-06 18:28:39 UNUSED OPTIONS
1 [verb] [4] 
2 [connect-retry] [2] [300] 
3 [resolv-retry] [60] 
12 [nobind] 
14 [pull-filter] [ignore] [peer-id] 
2020-06-06 18:28:39 Contacting <#HIDDEN#>:1194 via UDP
2020-06-06 18:28:39 Connecting to [vpn.<#HIDDEN#>]:1194 (<#HIDDEN#>) via UDPv4
2020-06-06 18:28:39 Tunnel Options:V4,dev-type tun,link-mtu 1558,tun-mtu 1500,proto UDPv4,comp-lzo,keydir 1,cipher AES-128-CBC,auth SHA1,keysize 128,tls-auth,key-method 2,tls-client
2020-06-06 18:28:39 Creds: UsernameEmpty/PasswordEmpty
2020-06-06 18:28:39 Peer Info:
IV_VER=3.6_git:master
IV_PLAT=android
IV_NCP=2
IV_TCPNL=1
IV_PROTO=2
IV_CIPHERS=AES-256-GCM:AES-128-GCM:CHACHA20-POLY1305
IV_LZO=1
IV_AUTO_SESS=1
IV_GUI_VER=de.blinkt.openvpn 0.7.15
IV_SSO=openurl,crtext
2020-06-06 18:28:39 VERIFY OK: depth=1, /CN=vpn.<#HIDDEN#>
2020-06-06 18:28:39 VERIFY OK: depth=0, /CN=vpn.<#HIDDEN#>
2020-06-06 18:28:39 SSL Handshake: CN=vpn.<#HIDDEN#>, TLSv1.3, cipher TLSv1.3 TLS_AES_256_GCM_SHA384, 2048 bit RSA
2020-06-06 18:28:39 Session is ACTIVE
2020-06-06 18:28:39 Sending PUSH_REQUEST to server...
2020-06-06 18:28:39 OPTIONS:
0 [route-gateway] [192.168.15.1] 
1 [topology] [subnet] 
2 [ping] [10] 
3 [ping-restart] [120] 
4 [route] [192.168.10.0] [255.255.255.0] 
5 [route] [192.168.20.0] [255.255.255.0] 
6 [dhcp-option] [DNS] [192.168.10.1] 
7 [dhcp-option] [DOMAIN] [<#HIDDEN#>] 
8 [ifconfig] [192.168.15.21] [255.255.255.0] 
9 [peer-id] [1] 
2020-06-06 18:28:39 PROTOCOL OPTIONS:
  cipher: AES-128-CBC
  digest: SHA1
  compress: LZO
  peer ID: 1
2020-06-06 18:28:39 exception parsing IPv4 route: [route] [192.168.10.40/24] [] [net_gateway]  : tun_prop_error: route is not canonical
2020-06-06 18:28:39 We should call this sessionvpn.<#HIDDEN#>
2020-06-06 18:28:39 Opening tun interface:
2020-06-06 18:28:39 Local IPv4: 192.168.15.21/24 IPv6: (not set) MTU: 1500
2020-06-06 18:28:39 DNS Server: 192.168.10.1, Domain: <#HIDDEN#>
2020-06-06 18:28:39 Routes: 192.168.10.0/24, 192.168.20.0/24 
2020-06-06 18:28:39 Routes excluded:  
2020-06-06 18:28:39 VpnService routes installed: 192.168.10.0/24, 192.168.20.0/24 
2020-06-06 18:28:39 Disallowed VPN apps: 
2020-06-06 18:28:39 Connected via tun
2020-06-06 18:28:39 LZO init swap=0 asym=1
2020-06-06 18:28:39 COMPRESSION_ENABLED: Asymmetric compression enabled.  Server may send compressed data.  This may be a potential security issue.
2020-06-06 18:28:39 Debug state info: CONNECTED  to WIFI , pause: userPause, shouldbeconnected: true, network: SHOULDBECONNECTED
```

**Comment by schwabe** [comment]
Yes. As I suspected your VPN routes overlap the local network explicitly by specifying exactly the same network.

**Comment by domhaas** [comment]
> Yes. As I suspected your VPN routes overlap the local network explicitly by specifying exactly the same network.

Yes, of course. As I want to access my home-network from remote via VPN, by the same IP-addresses I need this route.
As I want to have VPN always connected (cause using private DNS) I currently got the problem that when I'm on Home-Wifi, the local addresses are routed via VPN, because of the route.
So I've configured the VPN-Client, that it does a reconnect when the network changes, this works. The only thing missing is, that the route pushed from the server should be ignored when it's the same as the local address.
So when the method `NetworkUtils.getLocalNetworks()` would give me the correct `192.168.10.0/24` and not as it does currently `192.168.10.40/24` this should work.

**Comment by schwabe** [comment]
Yeah but the argument works actually against you as well. If you are on a hotspot that happens to have the same local network as the VPN pushes you want that to be overridden by default.

**Comment by domhaas** [comment]
Yes, sure. I can live with that, as I'm not using untrusted hotspots.
I'm going find out why I'm getting an ip-address and not the network.

**Comment by schwabe** [comment]
You try to workaround this issue by pushing a less specific route from your VPN server. E.g. 192.168.10.0/24 that should make the exclude route take prefererence over the VPN route if it is present.

**Comment by domhaas** [comment]
Thanks for your support @schwabe. I was a little bit playing around with getting the local network - as expected `NetworkUtils->getLocalNetworks()` does not return networks but local-ips, also with a wrong CIDR like 192.168.10.123/24
So I still think this is a bug, because I tried this on several Devices with completely different Android.
I think the only way to get the local network(s) is parsing the routes. But anyhow, I'm closing this issue and coming back if I got something to merge.

Thank's for the great work!
